File: 2026-06/2026-06-25/frontmatter_parser.py
def parse_frontmatter(s):
    regex = re.compile(r"(?<=\n)(?P<key>.+?): (?P<value>.+?)(?=\n)")
    matches = [item.groupdict() for item in re.finditer(regex, s)]
    dictionary = {}

    for match in matches:
        key = match['key']
        value = match['value']

        if value.isdigit():
            dictionary[key] = int(value)
        elif value.count(".") == 1 and value.replace(".", "").isdigit():
            dictionary[key] = float(value)
        elif value.lower() == "true":
            dictionary[key] = True
        elif value.lower() == "false":
            dictionary[key] = False
        else:
            dictionary[key] = value

    return dictionary

# --- TEST SUITE ---

tests_text = r'''
1. parse_frontmatter("---\ntitle: My Post\ndraft: false\nviews: 100\n---") should return { 'title': "My Post", 'draft': False, 'views': 100 }.
2. parse_frontmatter("---\nid: 6a174db57256a112f932195c\ntitle: My Book\nlocale: en\nwordCount: 10000\npublished: false\n---") should return { 'id': "6a174db57256a112f932195c", 'title': "My Book", 'locale': "en", 'wordCount': 10000, 'published': False }.
3. parse_frontmatter("---\nversion: 1.0.0\nurl: https://example.com\nprivate: true\n---") should return { 'version': "1.0.0", 'url': "https://example.com", 'private': True }.
4. parse_frontmatter("---\nrating: 4.5\nprice: 9.99\n---") should return { 'rating': 4.5, 'price': 9.99 }.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 06 June/06-25/frontmatterParser.js
function parseFrontmatter(str) {
  const matches = [...str.matchAll(/(?<=\n)(?<key>.+?): (?<value>.+?)(?=\n)/g)];
  const object = {};

  for (const match of matches) {
    let { key, value } = match.groups;

    if (!isNaN(value)) {
      object[key] = Number(value);
    } else if (value === "true") {
      object[key] = true;
    } else if (value === "false") {
      object[key] = false;
    } else {
      object[key] = value;
    }
  }

  return object;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. parseFrontmatter("---\ntitle: My Post\ndraft: false\nviews: 100\n---") should return { title: "My Post", draft: false, views: 100 }.
2. parseFrontmatter("---\nid: 6a174db57256a112f932195c\ntitle: My Book\nlocale: en\nwordCount: 10000\npublished: false\n---") should return { id: "6a174db57256a112f932195c", title: "My Book", locale: "en", wordCount: 10000, published: false }.
3. parseFrontmatter("---\nversion: 1.0.0\nurl: https://example.com\nprivate: true\n---") should return { version: "1.0.0", url: "https://example.com", private: true }.
4. parseFrontmatter("---\nrating: 4.5\nprice: 9.99\n---") should return { rating: 4.5, price: 9.99 }.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);